### DISCUSSION TITLE
Filesystem operations sometimes fail with an ebusy return code when EBS volumes are used for storage

### DISCUSSION THREAD

**Discussion by daveofthedogs** [rabbitmq/rabbitmq-server#11047] [Other] [answered]
### Describe the bug

AWS EC2
OS: Red Hat Enterprise Linux release 9.3 (Plow)
RMQ version:  3.13.0
Erlang: Erlang/OTP 26 [erts-14.2.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit:ns]
3-node cluster with rabbitmq_peer_discovery_aws and configuration set to ec2 tags

Disk setup:
```
/dev/nvme4n1          /dev/ng4n1            vol036b8eb943a110702 Amazon Elastic Block Store               0x1        322.12  GB / 322.12  GB    512   B +  0 B   1.0
/dev/nvme5n1          /dev/ng5n1            vol0a1dae2a212380a0f Amazon Elastic Block Store               0x1        322.12  GB / 322.12  GB    512   B +  0 B   1.0
/dev/nvme6n1          /dev/ng6n1            vol0af5c106f5d9d1d35 Amazon Elastic Block Store               0x1        107.37  GB / 107.37  GB    512   B +  0 B   1.0
/dev/nvme7n1          /dev/ng7n1            vol04c76232a475b611e Amazon Elastic Block Store               0x1        107.37  GB / 107.37  GB    512   B +  0 B   1.0
```

Partitioning:
```
Filesystem     1K-blocks    Used Available Use% Mounted on
/dev/nvme6n1   104792064  764576 104027488   1% /var/lib/rabbitmq
/dev/nvme4n1   314419200 2226032 312193168   1% /var/lib/rabbitmq/mnesia/rabbit@ip-10-117-128-59/quorum
/dev/nvme5n1   314419200 2226032 312193168   1% /var/lib/rabbitmq/mnesia/rabbit@ip-10-117-128-59/stream
/dev/nvme7n1   104792064  770648 104021416   1% /var/log/rabbitmq
```

Mount options:
```
/dev/nvme6n1 on /var/lib/rabbitmq type xfs (rw,noatime,attr2,inode64,logbufs=8,logbsize=32k,sunit=8,swidth=8,noquota)
/dev/nvme4n1 on /var/lib/rabbitmq/mnesia/rabbit@ip-10-117-128-59/quorum type xfs (rw,noatime,attr2,inode64,logbufs=8,logbsize=32k,sunit=8,swidth=8,noquota)
/dev/nvme5n1 on /var/lib/rabbitmq/mnesia/rabbit@ip-10-117-128-59/stream type xfs (rw,noatime,attr2,inode64,logbufs=8,logbsize=32k,sunit=8,swidth=8,noquota)
/dev/nvme7n1 on /var/log/rabbitmq type xfs (rw,noatime,attr2,inode64,logbufs=8,logbsize=32k,sunit=8,swidth=8,noquota)
```

Permissions:
```
      128      0 drwxr-xr-x   3 rabbitmq rabbitmq       42 Feb 22 17:58 /var/lib/rabbitmq
      131      0 drwxrwxr-x   4 rabbitmq rabbitmq      128 Apr 19 15:24 /var/lib/rabbitmq/mnesia
 16777344      0 drwxr-x---   4 rabbitmq rabbitmq       34 Apr 19 15:24 /var/lib/rabbitmq/mnesia/rabbit@ip-10-117-128-59
      128      0 drwxr-x---   2 rabbitmq rabbitmq        6 Apr 19 15:24 /var/lib/rabbitmq/mnesia/rabbit@ip-10-117-128-59/quorum
      128      0 drwxr-x---   2 rabbitmq rabbitmq        6 Apr 19 14:27 /var/lib/rabbitmq/mnesia/rabbit@ip-10-117-128-59/stream
 50331777      0 drwxr-x---   2 rabbitmq rabbitmq        6 Apr 19 15:24 /var/lib/rabbitmq/mnesia/rabbit@ip-10-117-128-59-plugins-expand
```

Observed Error:
```
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0> BOOT FAILED
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0> ===========
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0> Exception during startup:
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0>
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0> error:{badmatch,{error,{"/var/lib/rabbitmq/mnesia/rabbit@ip-10-117-128-59/quorum",ebusy}}}
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0>
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0>     rabbit_mnesia:wipe/0, line 263
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0>     rabbit_db:reset/0, line 140
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0>     rabbit_db_cluster:join/2, line 137
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0>     rabbit_peer_discovery:join_selected_node_locked/2, line 913
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0>     rabbit_peer_discovery:join_selected_node/3, line 883
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0>     rabbit_peer_discovery:sync_desired_cluster/3, line 193
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0>     rabbit_db:init/0, line 65
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0>     rabbit_boot_steps:-run_step/2-lc$^0/1-0-/2, line 51
2024-04-19 15:03:09.359944+00:00 [error] <0.249.0>
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>   crasher:
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     initial call: application_master:init/4
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     pid: <0.248.0>
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     registered_name: []
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     exception exit: {{badmatch,{error,{"/var/lib/rabbitmq/mnesia/rabbit@ip-10-117-128-59/quorum",
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>                                        ebusy}}},
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>                      {rabbit,start,[normal,[]]}}
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>       in function  application_master:init/4 (application_master.erl, line 142)
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     ancestors: [<0.247.0>]
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     message_queue_len: 1
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     messages: [{'EXIT',<0.249.0>,normal}]
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     links: [<0.247.0>,<0.44.0>]
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     dictionary: []
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     trap_exit: true
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     status: running
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     heap_size: 2586
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     stack_size: 28
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>     reductions: 177
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>   neighbours:
2024-04-19 15:03:10.360968+00:00 [error] <0.248.0>
2024-04-19 15:03:10.366786+00:00 [notice] <0.44.0> Application rabbit exited with reason: {{badmatch,{error,{"/var/lib/rabbitmq/mnesia/rabbit@ip-10-117-128-59/quorum",ebusy}}},{rabbit,s
tart,[normal,[]]}}
```

Error does not occur on 100% of servers .

### Reproduction steps

1. Build rhel9 server with separate disks for /var/log/rabbitmq, /var/lib/rabbitmq, /var/lib/rabbitmq/mnesia/<nodename>/quorum, and var/lib/rabbitmq/mnesia/<nodename>/stream
2. Create rabbitmq gid and rabbitmq uid
3. Set ownership of directories to rabbitmq:rabbitmq, 775
4. Install and start rabbitmq with configs utilizing quorum queues
...

### Expected behavior

Expected behavior is a good startup with rabbitmq being able to create directories and files under /var/lib/rabbitmq/mnesia/<nodename>/quorum

**Comment by kjnilsson** [comment]
What kind of EBS volumes are you using?

**Reply by michaelklishin** [reply]
`Amazon Elastic Block Store` suggests these are EBS volumes. They are not directly attached.

**Reply by lukebakken** [reply]
I think @kjnilsson is wondering if they are using HDD EBS volumes - https://docs.aws.amazon.com/ebs/latest/userguide/ebs-volume-types.html

**Comment by michaelklishin** [comment]
@daveofthedogs you cannot claim that an `ebusy` return code from a filesystem operation is a bug in RabbitMQ (or its Erlang runtime), so while we would be interested in learning about what kind of conditions can result in such exceptions, this is an environment-specific problem way outside of RabbitMQ control.

**Comment by michaelklishin** [comment]
Our [Production Checklist guide](https://www.rabbitmq.com/docs/production-checklist#storage) explicitly suggests against using network-attached storage where possible. It is certainly possible to use local storage on AWS, including SSDs (that quorum queues, Khepri and in particular streams will greatly benefit from).

This question should be directed to AWS support, and these nodes should use VM-local storage, as recommended in the docs.

**Reply by daveofthedogs** [reply]
@michaelklishin ,
I have over 800 RMQ servers I am finally attempting to migrated from mirrored to quorum queues.  It is most probably going to be cost-prohibitive to supply locally-attached storage for all of these.

The documentation states this:
```
Network-attached Storage (NAS)
Network-attached storage (NAS) can be used for RabbitMQ node data directories, provided that the NAS volume

It offers low I/O latency
It can guarantee no significant latency spikes (for example, due to sharing with other I/O-heavy services)
Quorum queues, streams, and other RabbitMQ features will benefit from fast local SSD and NVMe storage. When possible, prefer local storage to NAS.
```

Note, the storage I provided is:
1. Dedicated volumes for all, quorum, streams (not currently used), and mnesia.
2. Thrse are SDDs (io2)
3. These are NVMe at a rate of iops from 1000 to 3000 (testing going higher now).

What are the actual latency and throughput requirements of the disks? I have seen articles by Jack Vanlightly regarding message throughput and comparing hdd vs sdd and mixed loads vs single-types of workloads.  These have shown SDD/Quorum workloads reaching 15ms latencies.  I cannot even start RMQ 

 I am wondering how VMWare/Pivotal handles this situation since they have moved from traditional deployments to Kubernetes, which is also not locally-attached storage under most circumstances.

Below this post, I am including fio write stats running against a working install and non-working.  Both are exactly the same server builds with the exception of the availability zone of the servers and ebs volumes.

I am opening a ticket with AWS, but I  highly suspect that locally attached storage will be cost prohibitive. 

---

Tests io2 ebs @ 1000 iops

<details><summary>Working server (us-east-1a)</summary>

```
fio_test_file: (g=0): rw=randwrite, bs=(R) 16.0KiB-16.0KiB, (W) 16.0KiB-16.0KiB, (T) 16.0KiB-16.0KiB, ioengine=psync, iodepth=1
...
fio-3.35
Starting 16 processes

fio_test_file: (groupid=0, jobs=16): err= 0: pid=590063: Sun Apr 21 19:04:59 2024
  write: IOPS=1005, BW=15.7MiB/s (16.5MB/s)(2828MiB/180019msec); 0 zone resets
    clat (usec): min=282, max=28053, avg=15910.44, stdev=3660.10
     lat (usec): min=282, max=28054, avg=15910.91, stdev=3660.11
    clat percentiles (usec):
     |  1.00th=[ 5997],  5.00th=[ 8029], 10.00th=[10028], 20.00th=[15926],
     | 30.00th=[15926], 40.00th=[16057], 50.00th=[16057], 60.00th=[16057],
     | 70.00th=[16057], 80.00th=[17957], 90.00th=[20055], 95.00th=[21890],
     | 99.00th=[22938], 99.50th=[23987], 99.90th=[26084], 99.95th=[26084],
     | 99.99th=[27132]
   bw (  KiB/s): min=13120, max=48000, per=100.00%, avg=16095.82, stdev=112.40, samples=5744
   iops        : min=  820, max= 3000, avg=1005.99, stdev= 7.03, samples=5744
  lat (usec)   : 500=0.56%, 750=0.01%
  lat (msec)   : 2=0.01%, 4=0.01%, 10=9.46%, 20=78.67%, 50=11.30%
  cpu          : usr=0.04%, sys=0.06%, ctx=181055, majf=0, minf=155
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=0,181004,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=1

Run status group 0 (all jobs):
  WRITE: bw=15.7MiB/s (16.5MB/s), 15.7MiB/s-15.7MiB/s (16.5MB/s-16.5MB/s), io=2828MiB (2966MB), run=180019-180019msec

Disk stats (read/write):
  nvme4n1: ios=0/180889, merge=0/0, ticks=0/2874979, in_queue=2874978, util=99.99%
```
</details>

<details><summary>Not working (us-east-1b)</summary>

```
fio_test_file: (g=0): rw=randwrite, bs=(R) 16.0KiB-16.0KiB, (W) 16.0KiB-16.0KiB, (T) 16.0KiB-16.0KiB, ioengine=psync, iodepth=1
...
fio-3.35
Starting 16 processes

fio_test_file: (groupid=0, jobs=16): err= 0: pid=602333: Sun Apr 21 19:04:59 2024
  write: IOPS=1005, BW=15.7MiB/s (16.5MB/s)(2828MiB/180016msec); 0 zone resets
    clat (usec): min=283, max=28058, avg=15912.40, stdev=3597.98
     lat (usec): min=284, max=28059, avg=15912.93, stdev=3597.99
    clat percentiles (usec):
     |  1.00th=[ 5997],  5.00th=[ 8029], 10.00th=[10028], 20.00th=[15926],
     | 30.00th=[15926], 40.00th=[16057], 50.00th=[16057], 60.00th=[16057],
     | 70.00th=[16057], 80.00th=[17957], 90.00th=[20055], 95.00th=[21103],
     | 99.00th=[23987], 99.50th=[23987], 99.90th=[26084], 99.95th=[26084],
     | 99.99th=[27919]
   bw (  KiB/s): min=12512, max=48000, per=100.00%, avg=16094.22, stdev=114.79, samples=5744
   iops        : min=  782, max= 3000, avg=1005.89, stdev= 7.17, samples=5744
  lat (usec)   : 500=0.56%, 750=0.01%
  lat (msec)   : 2=0.01%, 4=0.01%, 10=8.87%, 20=79.73%, 50=10.83%
  cpu          : usr=0.02%, sys=0.13%, ctx=181033, majf=0, minf=182
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=0,180980,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=1

Run status group 0 (all jobs):
  WRITE: bw=15.7MiB/s (16.5MB/s), 15.7MiB/s-15.7MiB/s (16.5MB/s-16.5MB/s), io=2828MiB (2965MB), run=180016-180016msec

Disk stats (read/write):
  nvme5n1: ios=0/181499, merge=0/5, ticks=0/2883090, in_queue=2883089, util=99.99%
```
</details>

**Reply by michaelklishin** [reply]
Disks returning `ebusy` often is not a problem our team sees elsewhere, although one specific Kubernetes provider does exhibit fairly unique I/O failures some of which future versions of Ra and RabbitMQ will deal better with.

Disk I/O errors preventing nodes from starting is not a commonly reported issue, and the recommendation would generally be "use volumes that do not exhibit this behavior". RabbitMQ nodes can and usually will produce a lot of I/O reads on boot, the write I/O should be fairly limited up to the point when client listeners are enabled.

Maybe an ongoing routing of disk I/O chaos testing and improvements would help with these EBS devices, or maybe not, they are not the norm across the (very diverse) customer and user base our team interacts with.

**Reply by lukebakken** [reply, marked as answer]
FWIW, this discussion is literally the first time I have seen `ebusy` in a RabbitMQ environment. Clearly there is something odd in your specific case.

@daveofthedogs any time you include the output of a command, it is _essential_ that you also include the command invocation itself. We have no idea how you are running `fio`.

> Build rhel9 server with separate disks for /var/log/rabbitmq, /var/lib/rabbitmq, /var/lib/rabbitmq/mnesia//quorum, and var/lib/rabbitmq/mnesia//stream

Do you see this issue when you do _NOT_ separate out disks like this?

**Reply by daveofthedogs** [reply]
Sorry. I did not see the last part of your query.  I will modify my terraform to only build with a large, fast root volume.

**Reply by daveofthedogs** [reply]
@lukebakken , it looks like the simplified file system layout has helped.  The concern I had/have was after reading @Vanlightly's articled about separating the WAL and segemt files i/o load.
https://www.rabbitmq.com/blog/2020/04/21/quorum-queues-and-why-disks-matter
https://www.rabbitmq.com/blog/2020/06/21/cluster-sizing-case-study-quorum-queues-part-1
and
https://www.cloudamqp.com/blog/rabbitmq-directories-and-files-part-2-data-directory.html
"Moving into the node directory, default@localhost, the msg_stores folder houses all the classic queues related data, quorum and stream folders contain the quorum and stream queues-related data respectively.
"

Perhaps my issue was specifically mounting a volume for the quorum? I was trying to isolate the log and data volumes.

Current layout:
```
[root@ip-10-117-128-112 ~]# df -h
Filesystem      Size  Used Avail Use% Mounted on
devtmpfs        4.0M     0  4.0M   0% /dev
tmpfs           7.4G     0  7.4G   0% /dev/shm
tmpfs           3.0G   11M  3.0G   1% /run
/dev/nvme0n1p4  100G  5.5G   94G   6% /
tmpfs           512M  280K  512M   1% /tmp
/dev/nvme0n1p3  495M  275M  220M  56% /boot
/dev/nvme2n1    8.0G  1.8G  6.2G  23% /opt
/dev/nvme0n1p2  200M   20K  200M   1% /efi
/dev/nvme1n1     10G  358M  9.6G   4% /var/log
/dev/nvme3n1    2.0G  236M  1.8G  12% /var/log/audit
/dev/nvme4n1    300G  2.2G  298G   1% /var/lib/rabbitmq <<<<<<<
/dev/nvme5n1    100G  747M  100G   1% /var/log/rabbitmq <<<<<<<
tmpfs           1.5G  4.0K  1.5G   1% /run/user/35638
```

This worked 100 %. Not sure if I should take this as a win and just publish these settings or try to figure out ther issue with the separate quorum volume. ¯\_(ツ)_/¯

Thanks for the followup.

**Reply by michaelklishin** [reply]
@daveofthedogs those blog posts are four years old and Jack is no longer on our team. Tagging him or other team members directly here is completely unnecessary.

When RabbitMQ or its Raft implementation performs file I/O, it does not know what volume the target directory is on. It should not have to. Most people do not store QQ data directory on a separate volume, which may explain why we haven't see this behavior before.

But when someone does use a separate volume, it is always to move QQ and stream data directories devices that offer higher write throughput. Using N EBS volumes over one won't necessarily make a lot of difference.

Can this behavior be reproduced with a single EBS volume used for all of node's data?

**Comment by mkuratczyk** [comment]
This error has nothing to do with performance but rather indeed with volume mounting. It can be easily triggered by mounting a volume in the `quorum/` folder (or any other folder for that matter):
1. Mount a different folder, something like: `sudo mount -o size=1G -t tmpfs none $PWD/quorum`
2. Start the node
3. Try to reset the node:
```
$ rabbitmqctl stop_app
Stopping rabbit application on node rabbit@carrot ...
$ rabbitmqctl reset
Resetting node rabbit@carrot ...
Error:
{{:badmatch, {:error, {~c"/mnt/data/rabbit@carrot/mnesia/rabbit@carrot/quorum", :ebusy}}}, [{:rabbit_mnesia, :wipe, 0, [file: ~c"rabbit_mnesia.erl", line: 263]}, {:rabbit_db, :reset, 0, [file: ~c"rabbit_db.erl", line: 140]}]}
```

The same will happen if trying to remove this folder from the operating system level (which is effectively the same thing that RabbitMQ does):
```
> rm -rf quorum/
rm: cannot remove 'quorum/': Device or resource busy
```

`rabbit_file:recursive_delete/1` should probably just ignore this error and carry on

**Reply by daveofthedogs** [reply]
@mkuratczyk Is it a bad practice to have a dedicated volume for quorum?  I figured it would give a more optimal i/o profile for the WAL and other data.

**Reply by mkuratczyk** [reply]
Have I said anything that would suggest this is a bad practice?

A vast majority of users doesn't do that, which would explain why we haven't seen this before. I'd expect this separation to have some performance benefit, although I'd be surprised if it was significant under typical workloads. Given you are already testing this, please share what difference you see between using and not using a separate volume.

**Reply by lukebakken** [reply]
cc @mkuratczyk - https://github.com/rabbitmq/rabbitmq-server/pull/11066

**Comment by daveofthedogs** [comment]
Closing this discussion, as the answer was to not mount the quorum directory separately,  Team reproduced the error and suggested a fix for it.

**Reply by lukebakken** [reply]
Thanks for following up.

**Comment by michaelklishin** [comment]
According to [`glibc` docs](https://www.gnu.org/software/libc/manual/html_node/Deleting-Files.html), `ebusy` means that a file cannot be unlinked (deleted) immediately because it is used in some shape or form according to the kernel.

So looks like we must retry up to N times with a delay in between (of, say, 5ms?) and it (in theory) should only affect file deletes, for which we usually can afford to retry N times.

**Reply by michaelklishin** [reply]
In addition, I cannot find anything specific about EBS or multi-volume scenarios but I have a hypothesis that when there are multiple mounts and they overlap, the Linux kernel will use a different filesystem operation locking strategy (a more coarse-grained one), which may explain why the following sequence of operations may run into an `ebusy`:

1. A file read (or write)
2. An application-level decision that the file now must be deleted
3. An attempted delete

**Reply by michaelklishin** [reply]
@bakkenl has clarified what I'm missing in #11066: if you have a root node data directory on volume A, and one of its subdirectories on volume B, deleting all subdirectories from A, in particular B, won't be possible because a portion of the path is shared, and so are the associated kernel locks.

So ignoring the errors might help the node start but this configuration is going to cause problems one way or another, eventually.

**Comment by michaelklishin** [comment]
@bakkenl has clarified what I'm missing in #11066: if you have a root node data directory on volume A, and one of its subdirectories on volume B, deleting all subdirectories from A, in particular B, won't be possible because a portion of the path is shared, and so are the associated kernel locks.

An alternative multi-volume solution might look like this:

 * Place node data dir on volume A, say, at `/mnt/rabbitmq/base` or something like that
 * Mount a QQ subdirectory at a non-overlapping path, say, at `/mnt/rabbitmq/quorum_queues` and override that specific location

But it should be at least mentioned that this kind of setup is rare and we don't test or document it enough. And that for two (slow compared to local SSDs) EBS volumes it may or may not yield any practical improvements in I/O throughput or latency.

**Comment by michaelklishin** [comment]
Besides the suggestion to avoiding overlapping filesystem volumes, https://github.com/rabbitmq/rabbitmq-server/pull/11066 should at least avoid the exception under such conditions.
